--- a/libfreerdp/gdi/region.c
+++ b/libfreerdp/gdi/region.c
@@ -553,9 +553,11 @@ INLINE BOOL gdi_InvalidateRegion(HGDI_DC hdc, INT32 x, INT32 y, INT32 w,
 
 	if ((hdc->hwnd->ninvalid + 1) > hdc->hwnd->count)
 	{
-		int new_cnt;
+		size_t new_cnt;
 		HGDI_RGN new_rgn;
 		new_cnt = hdc->hwnd->count * 2;
+		if (new_cnt > UINT32_MAX)
+			return FALSE;
 		new_rgn = (HGDI_RGN) realloc(cinvalid, sizeof(GDI_RGN) * new_cnt);
 
 		if (!new_rgn)

