Author: Mike Gabriel <mike.gabriel@das-netzwerkteam.de>
Description: Spelling fix(es)

--- a/libfreerdp/codec/color.c
+++ b/libfreerdp/codec/color.c
@@ -1555,7 +1555,7 @@
 
 			if (xorBpp == 8 && !palette)
 			{
-				WLog_ERR(TAG, "null palette in convertion from %d bpp to %d bpp",
+				WLog_ERR(TAG, "null palette in conversion from %d bpp to %d bpp",
 						 xorBpp, dstBitsPerPixel);
 				return -1;
 			}
