Description: Spelling fix in log message.
Author: Mike Gabriel <mike.gabriel@das-netzwerkteam.de>

--- a/client/X11/xf_disp.c
+++ b/client/X11/xf_disp.c
@@ -409,7 +409,7 @@
 	if (settings->Fullscreen)
 		return CHANNEL_RC_OK;
 
-	WLog_DBG(TAG, "DisplayControlCapsPdu: setting the window as resizeable");
+	WLog_DBG(TAG, "DisplayControlCapsPdu: setting the window as resizable");
 	return xf_disp_set_window_resizable(xfDisp) ? CHANNEL_RC_OK : CHANNEL_RC_NO_MEMORY;
 }
 
